smokedduck/operators.py

```diff
@@ -0,0 +1,356 @@
+from abc import ABC, abstractmethod
+
+
+def get_op(op_str, query_id, parent_join_cond):
+    op, op_id = op_str.rsplit("_", 1)
+    if op == 'LIMIT':
+        return Limit(query_id, op_id, parent_join_cond)
+    elif op == 'FILTER':
+        return Filter(query_id, op_id, parent_join_cond)
+    elif op == 'ORDER_BY':
+        return OrderBy(query_id, op_id, parent_join_cond)
+    elif op == 'PROJECTION':
+        return Projection(query_id, op_id, parent_join_cond)
+    elif op == 'SEQ_SCAN':
+        return TableScan(query_id, op_id, parent_join_cond)
+    elif op == 'HASH_GROUP_BY':
+        return GroupBy(query_id, op_id, parent_join_cond)
+    elif op == 'PERFECT_HASH_GROUP_BY':
+        return PerfectGroupBy(query_id, op_id, parent_join_cond)
+    elif op == 'BLOCKWISE_NL_JOIN':
+        return BlockwiseNLJoin(query_id, op_id, parent_join_cond)
```

Comment:
similar to the current logic for BlockwiseNLJoin

Reply:
I've been going back and forth on if we should keep them separate or bring them together. What I like about keeping these separate is that, should a single operator get refactored, we wouldn't have to split things apart later. It also makes it a bit easier to find where one needs to make changes: everything relevant to building a query for a specific operator will be in that operator.

Do you feel strongly about merging them together to make the code DRYer? I could definitely be swayed if you do.

Reply:
I dont feel strongly about merging them, just thought it would be easier to maintain and fix, since the current logic for NestedLoopJoin and CrossProduct is not correct and the correct version is similar to BlockwiseNLJoin logic.